### ISSUE TITLE
AttributeError: module 'spinalcordtoolbox' has no attribute 'image'

### ISSUE THREAD

**Issue by jcohenadad** [spinalcordtoolbox/spinalcordtoolbox#3972]
### Description

A user from ETH contacted me to help with mouse MRIs. I suggested to use `sct_deepseg` with the Uqueensland mice model. However, when trying with the data the user sent me, I ran into the following bug (reproduced with the testing data):

<details><summary>Terminal output</summary>

```console
julien-macbook:~/code/spinalcordtoolbox/data/sct_testing_data/t2s $ sct_deepseg -i t2s.nii.gz -task seg_mice_sc

--
Spinal Cord Toolbox (git-master-d17e52f90a971cb778d93d88306b102254e9ba2f)

sct_deepseg -i t2s.nii.gz -task seg_mice_sc
--

2022-12-05 14:32:43.377 | WARNING  | ivadomed.inference:segment_volume:403 - fname_roi has not been specified, then the entire volume is processed.
2022-12-05 14:32:44.107 | INFO     | ivadomed.inference:segment_volume:453 - Loaded 9 axial slices.
2022-12-05 14:32:44.112 | INFO     | ivadomed.utils:define_device:157 - Cuda is not available.
2022-12-05 14:32:44.112 | INFO     | ivadomed.utils:define_device:158 - Working on cpu.
2022-12-05 14:32:44.112 | DEBUG    | ivadomed.inference:get_preds:71 - PyTorch model detected at: /Users/julien/code/spinalcordtoolbox/data/deepseg_models/mice_uqueensland_sc/mice_uqueensland_sc.pt
2022-12-05 14:32:44.112 | DEBUG    | ivadomed.inference:get_preds:72 - Loading model from: /Users/julien/code/spinalcordtoolbox/data/deepseg_models/mice_uqueensland_sc/mice_uqueensland_sc.pt
/Users/julien/code/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:656: SourceChangeWarning: source code of class 'ivadomed.models.Unet' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/Users/julien/code/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:656: SourceChangeWarning: source code of class 'ivadomed.models.Encoder' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/Users/julien/code/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:656: SourceChangeWarning: source code of class 'torch.nn.modules.container.ModuleList' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/Users/julien/code/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:656: SourceChangeWarning: source code of class 'ivadomed.models.DownConv' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/Users/julien/code/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:656: SourceChangeWarning: source code of class 'torch.nn.modules.conv.Conv2d' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/Users/julien/code/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:656: SourceChangeWarning: source code of class 'torch.nn.modules.batchnorm.BatchNorm2d' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/Users/julien/code/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:656: SourceChangeWarning: source code of class 'torch.nn.modules.dropout.Dropout2d' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/Users/julien/code/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:656: SourceChangeWarning: source code of class 'torch.nn.modules.pooling.MaxPool2d' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/Users/julien/code/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:656: SourceChangeWarning: source code of class 'ivadomed.models.Decoder' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/Users/julien/code/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/torch/serialization.py:656: SourceChangeWarning: source code of class 'ivadomed.models.UpConv' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
2022-12-05 14:32:44.142 | DEBUG    | ivadomed.inference:get_preds:75 - Evaluating model: /Users/julien/code/spinalcordtoolbox/data/deepseg_models/mice_uqueensland_sc/mice_uqueensland_sc.pt
2022-12-05 14:32:44.934 | DEBUG    | ivadomed.inference:get_preds:92 - Sending predictions to CPU
2022-12-05 14:32:44.958 | INFO     | ivadomed.utils:define_device:157 - Cuda is not available.
2022-12-05 14:32:44.958 | INFO     | ivadomed.utils:define_device:158 - Working on cpu.
2022-12-05 14:32:44.959 | DEBUG    | ivadomed.inference:get_preds:71 - PyTorch model detected at: /Users/julien/code/spinalcordtoolbox/data/deepseg_models/mice_uqueensland_sc/mice_uqueensland_sc.pt
2022-12-05 14:32:44.959 | DEBUG    | ivadomed.inference:get_preds:72 - Loading model from: /Users/julien/code/spinalcordtoolbox/data/deepseg_models/mice_uqueensland_sc/mice_uqueensland_sc.pt
2022-12-05 14:32:44.983 | DEBUG    | ivadomed.inference:get_preds:75 - Evaluating model: /Users/julien/code/spinalcordtoolbox/data/deepseg_models/mice_uqueensland_sc/mice_uqueensland_sc.pt
2022-12-05 14:32:45.106 | DEBUG    | ivadomed.inference:get_preds:92 - Sending predictions to CPU
Traceback (most recent call last):
  File "/Users/julien/code/spinalcordtoolbox/spinalcordtoolbox/scripts/sct_deepseg.py", line 253, in <module>
    main(sys.argv[1:])
  File "/Users/julien/code/spinalcordtoolbox/spinalcordtoolbox/scripts/sct_deepseg.py", line 234, in main
    fname_seg = ''.join([sct.image.splitext(input_filenames[0])[0], target + '.nii.gz'])
AttributeError: module 'spinalcordtoolbox' has no attribute 'image'
```

</details>

**Comment by joshuacwnewton** [comment]
This is a very simple syntax error, and can be easily fixed by making this change:

```python
# Before
import spinalcordtoolbox as sct
sct.image.splitext()

# After
from spinalcordtoolbox.image import splitext
splitext()
```

I will submit a fix ASAP (and perhaps even add the fix to the 5.8 release, given the urgency). But, for now, the collaborator should be able to run `sct_deepseg` by specifying `-o`:

```
sct_deepseg -i t2s.nii.gz -task seg_mice_sc -o t2s_seg.nii.gz
```

**Comment by joshuacwnewton** [comment]
For the bug to exist for that long seems farfetched to me... are there situations where the syntax `import sct; sct.<module>.<function>` would actually have worked? :thinking:

**Comment by joshuacwnewton** [comment]
Ah!! I think I have an idea of how this bug was introduced.

* `sct.utils.splitext` is odd, but acceptable syntax
* `sct.image.splitext` is _not_ acceptable syntax

The reason for this is because only the `utils` module is present in the namespace of `spinalcordtoolbox`:

```python
>>> import spinalcordtoolbox as sct
>>> dir(sct)
['__builtins__', '__cached__', '__data_dir__', '__deepseg_dir__', '__doc__', '__file__', '__loader__', 
 '__name__', '__package__', '__path__', '__sct_dir__', '__spec__', '__version__', 'utils']  # Why?!
```

Likely due to:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/812a8e00bd8862c3c9f2d099e24e0143a5337976/spinalcordtoolbox/__init__.py#L1

So, I believe the change in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/2949#discussion_r1040061573 introduced the error.

**Comment by mguaypaq** [comment]
I introduced the bug on July 15 this year: [diff](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3842/commits/3193fe20588f4b881a7282013bfee026ad4ebd3c#diff-c147295dd5c674b02253fe45ff93d2e7796d63131d93249b7d29506956c6a9ab).

You can see the weirdness in this little interaction:
```py3
>>> import spinalcordtoolbox as sct
>>> sct.image
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'spinalcordtoolbox' has no attribute 'image'
>>> from spinalcordtoolbox import image
>>> sct.image
<module 'spinalcordtoolbox.image' from '/home/mguaypaq/neuropoly/spinalcordtoolbox/spinalcordtoolbox/image.py'>
```

What's happening is this:
* `import spinalcordtoolbox as sct` creates a module object for `spinalcordtoolbox`, and binds it to the name `sct` in the current module (`sct_deepseg`).
* `from spinalcordtoolbox import image` sees that the `spinalcordtoolbox` module already exists (modules are only ever imported once, then they're cached), so it:
   * creates a module object for `image`,
   * binds it to the name `image` in the existing `spinalcordtoolbox` module object,
   * also binds it to the name `image` in the current module (`sct_deepseg`).

After that, as a side-effect, the `image` module object is also suddenly accessible as `sct.image` in the context of `sct_deepseg`.

I should probably check the other "unused" imports I cleaned up in #3842!

**Comment by joshuacwnewton** [comment]
> July 15 this year: [diff](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3842/commits/3193fe20588f4b881a7282013bfee026ad4ebd3c#diff-c147295dd5c674b02253fe45ff93d2e7796d63131d93249b7d29506956c6a9ab).

Oh! Thank goodness -- that means that only v5.7 and v5.8 are affected by this bug (since v5.7 was released on July 28th).

**Comment by mguaypaq** [comment]
The other potentially suspect import removals are these:
- [ ] [gui/sagittal.py](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3842/commits/3193fe20588f4b881a7282013bfee026ad4ebd3c#diff-98263975a806c478d51d740065c31e3beebc76cb50f8a0596d2ab2b55de8102c)
- [X] [scripts/isct_get_fractional_volume.py](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3842/commits/3193fe20588f4b881a7282013bfee026ad4ebd3c#diff-196dbd6ef2a99429897d15fd68811b9b95f836a546dd504cb118dd4eb6b6768a)
- [ ] [testing/create_test_data.py](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3842/commits/3193fe20588f4b881a7282013bfee026ad4ebd3c#diff-3032159954f3ebb765bfaf060f20e826123e2b7c1b7f4eee25585e246d89cb6f)

The second one is no longer even in SCT.

**Comment by joshuacwnewton** [comment]
This issue was re-reported by a [user on the forum](https://forum.spinalcordmri.org/t/error-in-seg-tumor-t2-task/1099/4). Perhaps we should consider backporting this fix to SCT v5.8?
